torch/_inductor/fx_passes/micro_pipeline_tp.py

```diff
@@ -0,0 +1,441 @@
+import operator
+from dataclasses import dataclass
+from typing import cast, List, Set, Tuple, Union
+
+import torch
+
+from ..pattern_matcher import (
+    CallFunction,
+    Ignored,
+    KeywordArg,
+    ListOf,
+    MULTIPLE,
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+
+aten = torch.ops.aten
+c10d = torch.ops._c10d_functional
+patterns = PatternMatcherPass()
+
+
+def _is_backward(graph: torch.fx.Graph) -> bool:
+    placeholders = []
+    for node in graph.nodes:
+        if node.op != "placeholder":
+            break
+        placeholders.append(node)
+    return not all(node.name.startswith("primal") for node in placeholders)
+
+
+def _compute_mm_arithmetic_intensity(M: int, N: int, K: int) -> float:
```

Comment:
I guess we are thinking of some heuristics to determine whether to apply micro-pipelining or not? This function ended up not getting used though.

Reply:
That's spot on! I was expecting to use arithmetic intensity as heuristic to exclude problem sizes not suitable for fusion. Will need to run more experiments to put it in place.